kernel/src/table_changes/resolve_dvs.rs

```diff
@@ -0,0 +1,422 @@
+            // `rm_dv` and `add_dv`, and they are initialized to the empty map by default.
+            // if no deletion vector is given.
+            //  rm_dv  = Treemap_d(0, 1)
+            //  add_dv = Treemap_d(1, 2)
+            //
+            // The result of this commit is:
+            // - row 0 is restored
+            // - row 1 is unchanged (previously deleted)
+            // - row 2 is newly deleted
+            // Thus for this commit we must generate `Treemap_s(0)` for the added rows, and
+            // `Treemap_s(2)` for deleted rows.
+            //
+            //  # Insertion Selection Treemap
+            //  The selection vector of added rows is calculated using set subtraction over deletion
+            //  treemaps `rm_dv - add_dv`. These rows went from set (deleted) in `rm_dv` to unset
+            //  (restored) in the `add_dv`. All other rows are either cancelled in set subtraction,
+            //  or were not selected in either treemap. Hence, they are not selected to
+            //  be in the set of changed rows. Applying this to our deletion treemaps:
+            //  rm_dv - add_dv =
+            //      Treemap_d(0, 1)
+            //    - Treemap_d(1, 2)
+            //    = Treemap_s(0)
+            //  The selection treemap shows that row 0 was inserted
+            //
+            //  # Deletion Selection Treemap
+            //  The selection vector of deleted rows is calculated using `add_dv - rm_dv`. These rows went
+            //  from unset (present) in `rm_dv` to set (deleted) in the `add_dv`. Once again, all
+            //  other rows are either cancelled in set subtraction, or were not set in either treemap.
+            //  Applying this to our deletion vectors:
+            //  add_dv - rm_dv =
+            //      Treemap_d(1, 2)
+            //    - Treemap_d(0, 1)
+            //    = Treemap_s(2)
+            //  The selection treemap shows that row 2 was deleted
+            //
+            //  # Conversion to Selection Vector
+            //  The selection treemap is converted to a selection vector by setting the bits:
+            //      Treemap_s(0) => [true]
+            //      Treemap_s(2) => [false, false, true]
+            //  All other rows are unselected (false).
+            let adds = &rm_dv - &add_dv;
+            let removes = add_dv - rm_dv;
+
+            let adds = (!adds.is_empty()).then_some(adds);
+            let removes = (!removes.is_empty()).then_some(removes);
+            (adds, removes)
+        }
+        (add_dv, None, CdfScanFileType::Add | CdfScanFileType::Cdc) => {
+            (Some(add_dv.unwrap_or_else(Default::default)), None)
+        }
+        (rm_dv, None, CdfScanFileType::Remove) => {
+            (None, Some(rm_dv.unwrap_or_else(Default::default)))
+        }
+    };
+    let treemap_to_bools = if scan_file.remove_dv.is_some() {
+        selection_treemap_to_bools
+    } else {
+        deletion_treemap_to_bools
+    };
```

Comment:
can we normalize above (i.e. everything to SV) and avoid this?

Reply:
Unfortunately not. Treemaps are sets, so constructing the inverse would require that we create a new "full" treemap containing every index (ie the number of rows in the data file). Then we take the full treemap and subtract out the selection treemap.

I think in most cases the set of deleted rows is much smaller than the set of all rows. So constructing such a treemap would be a lot more expensive.